crates/ruff_python_parser/src/parser/mod.rs

```diff
@@ -0,0 +1,4433 @@
+            while !self.at(TokenKind::EndOfFile) {
+                if self.at(TokenKind::Indent) {
+                    self.handle_unexpected_indentation(&mut body, "unexpected indentation");
+                    continue;
+                }
+                let (stmt, _) = self.parse_statement();
+                body.push(stmt);
+            }
+            ast::Mod::Module(ast::ModModule {
+                body,
+                // If the `source` only contains comments or empty spaces, return
+                // an empty range.
+                range: if is_src_empty {
+                    TextRange::default()
+                } else {
+                    TextRange::new(
+                        0.into(),
+                        self.source
+                            .len()
+                            .try_into()
+                            .expect("source length is  bigger than u32 max"),
+                    )
+                },
+            })
+        };
+
+        // After parsing, the `ctx` and `ctx_stack` should be empty.
+        // If it's not, you probably forgot to call `clear_ctx` somewhere.
+        assert!(self.ctx.is_empty() && self.ctx_stack.is_empty());
+
+        ParsedFile {
+            ast,
+            parse_errors: self.errors,
+        }
+    }
+
+    #[inline]
+    fn set_ctx(&mut self, ctx: ParserCtxFlags) {
+        self.ctx_stack.push(self.ctx);
+        self.ctx = ctx;
+    }
+
+    #[inline]
+    fn clear_ctx(&mut self, ctx: ParserCtxFlags) {
+        assert_eq!(self.ctx, ctx);
+        self.last_ctx = ctx;
+        if let Some(top) = self.ctx_stack.pop() {
+            self.ctx = top;
+        }
+    }
+
+    #[inline]
+    fn has_ctx(&self, ctx: ParserCtxFlags) -> bool {
+        self.ctx.intersects(ctx)
+    }
+
+    #[inline]
+    fn has_in_curr_or_parent_ctx(&self, ctx: ParserCtxFlags) -> bool {
+        self.has_ctx(ctx) || self.parent_ctx().intersects(ctx)
+    }
+
+    #[inline]
+    fn parent_ctx(&self) -> ParserCtxFlags {
+        self.ctx_stack
+            .last()
+            .copied()
+            .unwrap_or(ParserCtxFlags::empty())
+    }
+
+    fn next_token(&mut self) -> Spanned {
+        self.lexer
+            .next()
+            .map(|result| match result {
+                Ok(token) => token,
+                Err(lex_error) => {
+                    self.add_error(ParseErrorType::Lexical(lex_error.error), lex_error.location);
+
+                    // Return a `Invalid` token when encountering an error
+                    (Tok::Invalid, lex_error.location)
+                }
+            })
+            .unwrap_or((
+                Tok::EndOfFile,
+                TextRange::empty(
+                    self.source
+                        .len()
+                        .try_into()
+                        .expect("source length is bigger than u32 max"),
+                ),
+            ))
+    }
+
+    fn lookahead(&mut self, offset: usize) -> (TokenKind, TextRange) {
```

Comment:
At the moment, there's no way of knowing how much lookahead we need because of [this](https://github.com/astral-sh/ruff/blob/d4a4d056bc4db660500a89a3486ac5f7bf98dd50/crates/ruff_python_parser/src/parser/mod.rs#L1709) problem when parsing `WithItem`, but ideally it should be 1. I'm going to look how pyright handles this case.

Reply:
Ah interesting. Could we try parsing it as a parenthesized expression if the parser is at a `(` and and "fix up" the range if there's no following `as` keyword? That would avoid the need for the lookahead. It may require a special `parse_parenthesized_expression` method that returns the unparenthesized range as well.

Reply:
> Could we try parsing it as a parenthesized expression if the parser is at a ( and and "fix up" the range if there's no following as keyword?

Wouldn't work because if we have something like `(a as A, b)` and try to parse as a parenthesized expression, this example will be parsed as a tuple, resulting in an error message for the `as` keyword.